diff --git a/coregrind/m_syswrap/syswrap-linux.c b/coregrind/m_syswrap/syswrap-linux.c
index a1097c3..e90e30b 100644
--- a/coregrind/m_syswrap/syswrap-linux.c
+++ b/coregrind/m_syswrap/syswrap-linux.c
@@ -13227,6 +13227,34 @@ POST(sys_ioctl)
       break;
    }
 
+   case VKI_NTSYNC_CREATE_SEM:
+      if (!ML_(fd_allowed)(RES, "ntsync_create_sem", tid, True)) {
+         VG_(close)(RES);
+         SET_STATUS_Failure( VKI_EMFILE );
+      } else {
+         if (VG_(clo_track_fds))
+            ML_(record_fd_open_nameless) (tid, RES);
+      }
+      break;
+   case VKI_NTSYNC_CREATE_MUTEX:
+      if (!ML_(fd_allowed)(RES, "ntsync_create_mutex", tid, True)) {
+         VG_(close)(RES);
+         SET_STATUS_Failure( VKI_EMFILE );
+      } else {
+         if (VG_(clo_track_fds))
+            ML_(record_fd_open_nameless) (tid, RES);
+      }
+      break;
+   case VKI_NTSYNC_CREATE_EVENT:
+      if (!ML_(fd_allowed)(RES, "ntsync_create_event", tid, True)) {
+         VG_(close)(RES);
+         SET_STATUS_Failure( VKI_EMFILE );
+      } else {
+         if (VG_(clo_track_fds))
+            ML_(record_fd_open_nameless) (tid, RES);
+      }
+      break;
+
    default:
       /* EVIOC* are variable length and return size written on success */
       switch (ARG2 & ~(_VKI_IOC_SIZEMASK << _VKI_IOC_SIZESHIFT)) {
diff --git a/include/vki/vki-linux.h b/include/vki/vki-linux.h
index 00e228a..ec0bd9c 100644
--- a/include/vki/vki-linux.h
+++ b/include/vki/vki-linux.h
@@ -5544,6 +5544,25 @@ struct vki_open_how {
 #define VKI_CLOSE_RANGE_UNSHARE (1U << 1)
 #define VKI_CLOSE_RANGE_CLOEXEC (1U << 2)
 
+struct vki_ntsync_sem_args {
+    vki_uint32_t vki_count;
+    vki_uint32_t vki_max;
+};
+
+struct vki_ntsync_mutex_args {
+    vki_uint32_t vki_owner;
+    vki_uint32_t vki_count;
+};
+
+struct vki_ntsync_event_args {
+    vki_uint32_t vki_manual;
+    vki_uint32_t vki_signaled;
+};
+
+#define VKI_NTSYNC_CREATE_SEM     _VKI_IOR('N', 0x80, struct vki_ntsync_sem_args)
+#define VKI_NTSYNC_CREATE_MUTEX   _VKI_IOW('N', 0x84, struct vki_ntsync_mutex_args)
+#define VKI_NTSYNC_CREATE_EVENT   _VKI_IOW('N', 0x87, struct vki_ntsync_event_args)
+
 struct vki_cachestat_range {
     __vki_u64 off;
     __vki_u64 len;
