--- valgrind-3.11.0/coregrind/m_redir.c.0100~	2016-07-13 10:30:24.315653474 +0100
+++ valgrind-3.11.0/coregrind/m_redir.c	2016-07-13 10:33:19.368828703 +0100
@@ -1328,9 +1328,11 @@
       add_hardwired_spec(
          "ld-linux.so.2", "index",
          (Addr)&VG_(x86_linux_REDIR_FOR_index), mandatory);
+#if CI_HACK_strlen
       add_hardwired_spec(
          "ld-linux.so.2", "strlen",
          (Addr)&VG_(x86_linux_REDIR_FOR_strlen), mandatory);
+#endif
    }
 
 #  elif defined(VGP_amd64_linux)
@@ -1355,6 +1357,7 @@
       add_hardwired_spec(
          "ld-linux-x86-64.so.2", "index",
          (Addr)&VG_(amd64_linux_REDIR_FOR_index), NULL);
+#if CI_HACK_strlen
       add_hardwired_spec(
          "ld-linux-x86-64.so.2", "strlen",
          (Addr)&VG_(amd64_linux_REDIR_FOR_strlen),
@@ -1366,6 +1369,8 @@
          complain_about_stripped_glibc_ldso
 #        endif
       );   
+#endif
+
    }
 
 #  elif defined(VGP_ppc32_linux)
